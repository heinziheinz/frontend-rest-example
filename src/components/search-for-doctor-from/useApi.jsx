import React from 'react';

export const apiStates = {
    LOADING: 'LOADING',
    SUCCESS: 'SUCCESS',
    ERROR: 'ERROR',
};

const useApi = (url, options) => {
    const AbortController = window.AbortController;
    const myController = new AbortController();
    const mySignal = myController.signal;

    const [data, setData] = React.useState({
        state: apiStates.LOADING,
        error: '',
        data: [],
    });

    const setPartData = (partialData) => {
        console.log(partialData);
        setData({ ...data, ...partialData });
    }

    React.useEffect(() => {
        setPartData({
            state: apiStates.LOADING,
        });
        fetch(url, options)
            .then((response) => response.json())
            .then((data) => {
                setPartData({
                    state: apiStates.SUCCESS,
                    data
                });
            })
            .catch(() => {
                setPartData({
                    state: apiStates.ERROR,
                    error: 'fetch failed'
                });
            });
        return function cleanup() {
            myController.abort();
        };
    }, [options]);

    return data;
};

export default useApi;
