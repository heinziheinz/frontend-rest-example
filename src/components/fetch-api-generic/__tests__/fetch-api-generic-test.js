import React from 'react';
import PostList from './../fetch-api-generic';
import renderer from 'react-test-renderer';
import useApi from './../useApi';

jest.mock('./../useApi');

describe('PostList Snapshots', () => {
    it('loading renders correctly', () => {
        useApi.mockReturnValue({
            state: 'LOADING',
            error: '',
            data: [],
        });
        const tree = renderer.create(<PostList title="Test" />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('success renders correctly', () => {
        useApi.mockReturnValue({
            state: 'SUCCESS',
            error: '',
            data: [
                {
                    title: 'Hello',
                }, {
                    title: 'World',
                }
            ],
        });
        const tree = renderer.create(<PostList title="Test" />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('error renders correctly', () => {
        useApi.mockReturnValue({
            state: 'ERROR',
            error: 'General Error',
            data: [],
        });
        const tree = renderer.create(<PostList title="Test" />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});