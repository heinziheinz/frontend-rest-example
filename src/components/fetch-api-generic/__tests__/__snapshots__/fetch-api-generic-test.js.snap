// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostList Snapshots error renders correctly 1`] = `
<p>
  ERROR: 
  General Error
</p>
`;

exports[`PostList Snapshots loading renders correctly 1`] = `
<p>
  loading..
</p>
`;

exports[`PostList Snapshots success renders correctly 1`] = `
Array [
  <p>
    Data:
  </p>,
  <ul>
    <li>
      Hello
    </li>
    <li>
      World
    </li>
  </ul>,
]
`;
